libsolutil/Dominator.h

```diff
@@ -0,0 +1,377 @@
+		solAssert(!m_dfsIndexByVertexId.empty());
+		solAssert(!m_immediateDominators.empty());
+
+		std::vector<VId> dominators{};
+		solAssert(m_dfsIndexByVertexId.count(_vId));
+		// No one dominates the entry vertex and we consider self-dominance implicit
+		// i.e. all nodes already dominates itself.
+		if (m_dfsIndexByVertexId.at(_vId) == 0)
+			return dominators;
+
+		solAssert(m_dfsIndexByVertexId.count(_vId));
+		solAssert(m_dfsIndexByVertexId.at(_vId) < m_immediateDominators.size());
+		DfsIndex idomIdx = m_immediateDominators.at(m_dfsIndexByVertexId.at(_vId));
+		solAssert(idomIdx < m_immediateDominators.size());
+		while (idomIdx != 0)
+		{
+			solAssert(m_immediateDominators.at(idomIdx) < idomIdx);
+			dominators.emplace_back(m_verticesInDFSOrder.at(idomIdx));
+			idomIdx = m_immediateDominators[idomIdx];
+		}
+		// The loop above discovers the dominators in the reverse order
+		// i.e. from the given vertex upwards to the entry node (the root of the dominator tree).
+		// And the entry vertex always dominates all other vertices.
+		dominators.emplace_back(m_verticesInDFSOrder[0]);
+
+		return dominators;
+	}
+
+	/// Find all dominators of a node _v
+	/// @note for a vertex ``_v``, the _v’s inclusion in the set of dominators of ``_v`` is implicit.
+	std::vector<VId> dominatorsOf(V const& _v) const
+	{
+		return dominatorsOf(_v.id);
+	}
+
+private:
+	std::vector<DfsIndex> findDominators(V const& _entry)
+	{
+		// semi(w): The DFS index of the semidominator of ``w``.
+		std::vector<DfsIndex> semi{};
+		// parent(w): The index of the vertex which is the parent of ``w`` in the spanning
+		// tree generated by the DFS.
+		std::vector<DfsIndex> parent{};
+		// label(w): The index of the vertex ``w`` with the minimum semidominator in the path
+		// to its parent.
+		std::vector<DfsIndex> label{};
+		// predecessors(w): The set of vertices ``v`` such that (``v``, ``w``) is an edge of the graph.
+		std::vector<std::set<DfsIndex>> predecessors{};
+
+		// step 1
+		std::set<VId> visited;
+		// The number of vertices reached during the DFS.
+		// The vertices are indexed based on this number.
+		DfsIndex dfIdx = 0;
+		auto dfs = [&](V const& _v, auto _dfs) -> void {
+			auto [_, inserted] = visited.insert(_v.id);
+			if (!inserted)
+				return;
+			m_verticesInDFSOrder.emplace_back(_v.id);
+			m_dfsIndexByVertexId[_v.id] = dfIdx;
+			semi.emplace_back(dfIdx);
+			label.emplace_back(dfIdx);
+			// NOTE: Since each vertex is only visited once, incrementing dfIdx before visiting successors ensures
+			// it points to the next vertex to be visited (i.e., w).
+			// This is because we do not know the index of a vertex before visiting them.
+			dfIdx++;
+			ForEachSuccessor{}(_v, [&](V const& w) {
+				if (!visited.count(w.id))
+				{
+					// Here, ``dfIdx`` is the index of the vertex ``w`` in the DFS order.
+					if (dfIdx >= parent.size())
+						parent.resize(dfIdx + 1);
+					parent[dfIdx] = m_dfsIndexByVertexId[_v.id];
```

Comment:
Won't this leave garbage (or worse, `0`) as a value for the parent of the root vertex? This value is probably never used, but still it would be best to explicitly put something sane, like `std::numeric_limits<DfsIndex>::max()` in there.

Reply:
Looks like root's `parent` value does get used - it gets inserted into `ancestor` by `link()`. Fortunately it's at the very end of the main loop so nothing gets done with it.

In any case, one of my fixups includes https://github.com/ethereum/solidity/pull/14460#discussion_r1676293084 which also properly sets value for it. It's now set to `std::numeric_limits<DfsIndex>::max()` for the type, which is used in `ancestor` to mean "no parent".